src/tribler/core/components/restapi/rest/utils.py

```diff
@@ -66,3 +73,42 @@ def fix_unicode_array(arr):
             new_arr.append(item)
 
     return new_arr
+
+
+def shorten(s: Optional[str], width: int = 50, placeholder='[...]', cut_at_the_end: bool = True):
+    """ Shorten a string to a given width.
+        Example: shorten('hello world', 5) -> 'hello[...]'
+    """
+    if s and len(s) > width:
+        return f'{s[:width]}{placeholder}' if cut_at_the_end else f'{placeholder}{s[len(s) - width:]}'
+    return s
+
+
+def format_frames(frame: Optional[FrameType], file_width: int = 50, value_width: int = 100) -> Iterator[str]:
+    """ Format a stack trace."""
+    while frame:
+        filename = shorten(frame.f_code.co_filename, width=file_width, cut_at_the_end=False)
+        header = f"{filename}:{frame.f_lineno} {frame.f_code.co_name}"
+        local = ''
+        for key, value in list(frame.f_locals.items()):
+            value = shorten(repr(value), width=value_width)
+            local += f'\n\t{key} = {value}'
+        frame = frame.f_back
+        yield header + local
+
+
+def get_threads_info() -> List[Dict]:
```

Comment:
I'm a bit nervous about that function, as it is not thread-safe. The function call execution can potentially be very slow, especially because we calculate the repr of some random values, and some reprs can be huge, for example, for some JSON structures.

That means that the thread switch can happen during this calculation, and it is possible that the other thread will change the value, the repr of which we are calculating in the current thread, and it may not be thread-safe.

Even without repr rendering, the context switch can happen at any time when we traverse `frame = frame.f_back`, and it probably can lead to some surprising results.

So, it is important to avoid context switches while we are calculating the result. We can do it by setting `sys.setswitchinterval()` to a reasonably high value and then restoring it at the function end. But we should avoid the race condition with `main_thread_stack_tracking._get_main_thread_stack_info` that also calls `sys.setswitchinterval()` and can be executed in a parallel thread.

To avoid the race condition, we can introduce a context manager like:

```python
switch_interval_lock = Lock()

@contextmanager
def switch_interval(value):
    with switch_interval_lock:
        prev_value = sys.getswitchinterval(value)
        sys.setswitchinterval(value)
        try:
            yield
        finally:
            sys.setswitchinterval(prev_value)
```

And use it in all functions that currently call `sys.setswitchinterval`, like:

```python
def get_threads_info() -> List[Dict]:
    with switch_interval(ONE_SECOND):
        ...
```

Reply:
Isn't it too complicated for a minor feature in the debug tab?

Reply:
After some thinking, I agree that this improvement should be implemented. Thank you.